### PR TITLE
Enforce consumes interface in output modules

### PR DESCRIPTION
**Pull request by wmtan**
Output modules currently have direct access to EventPrincipal, RunPrincipal, and LuminosityBlockPrincipal. Calls to getByLabel() or getForOutput() functions of Principal bypass the consumes interface, so missing consumes calls will not be detected. This PR modifies output modules so that they do not access Principals directly. Rather, they use new classes EventForOutput, LuminosityBlockForOutput, and RunForOutput, similar to the Event, LuminosityBlock, and Run classes used by other modules. Furthermore, reading data is done only by getByToken(), so the consumes interface is enforced everywhere. getByToken() is the only get method supported by the new classes.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by wmtan** [comment]
@cmsbuild The three failing unit tests fail in the base IB. The failures have nothing to do with this PR.

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
+1

File: FWCore/Framework/interface/EventForOutput.h [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    EventAuxiliary const& aux_;
+    std::shared_ptr<LuminosityBlockForOutput const> const luminosityBlock_;
+
+    // gotBranchIDs_ must be mutable because it records all 'gets',
+    // which do not logically modify the PrincipalGetAdapter. gotBranchIDs_ is
+    // merely a cache reflecting what has been retreived from the
+    // Principal class.
+    typedef std::set<BranchID> BranchIDSet;
+    mutable BranchIDSet gotBranchIDs_;
+    void addToGotBranchIDs(Provenance const& prov) const;
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/EventForOutput.h]: We don't need to record the provenance for output modules.

File: FWCore/Framework/interface/TriggerResultsBasedEventSelector.h [hunk 2]
```diff
@@ -29,6 +31,13 @@ namespace edm {
     public:
       NamedEventSelector(std::string const& n, EventSelector const& s) :
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/TriggerResultsBasedEventSelector.h]: Is this constructor even needed anymore?

File: FWCore/Framework/src/EventForOutput.cc [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  bool
+  EventForOutput::getByToken(EDGetToken token, TypeID const& typeID, BasicHandle& result) const {
+    result.clear();
+    result = provRecorder_.getByToken_(typeID, PRODUCT_TYPE, token, moduleCallingContext_);
+    if (result.failedToGet()) {
+      return false;
+    }
+    addToGotBranchIDs(*result.provenance());
```
Comment (Dr15Jones) [inline FWCore/Framework/src/EventForOutput.cc]: We shouldn't need to do.

File: IOPool/Output/src/RootOutputFile.cc [hunk 4]
```diff
@@ -761,10 +756,74 @@ namespace edm {
     }
 
     if(productProvenanceVecPtr != nullptr) productProvenanceVecPtr->assign(provenanceToKeep.begin(), provenanceToKeep.end());
-    treePointers_[branchType]->fillTree();
+    treePointers_[InEvent]->fillTree();
     if(productProvenanceVecPtr != nullptr) productProvenanceVecPtr->clear();
   }
 
+  void RootOutputFile::fillBranches(LuminosityBlockForOutput const& lumi) {
```
Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputFile.cc]: In order to share the code, would it be better to either make this function templated or to create a base class for all `*ForOutput` classes?

**Comment by Dr15Jones** [comment]
Done with my comments

File: FWCore/Framework/interface/EventForOutput.h (hunk 1 above)
Comment (wmtan) [inline FWCore/Framework/interface/EventForOutput.h]: @Dr15Jones Yes, you are correct. This will be removed.

File: IOPool/Output/src/RootOutputFile.cc (hunk 4 above)
Comment (wmtan) [inline IOPool/Output/src/RootOutputFile.cc]: @Dr15Jones I agree.  I will create a base class.

File: FWCore/Framework/interface/TriggerResultsBasedEventSelector.h (hunk 2 above)
Comment (wmtan) [inline FWCore/Framework/interface/TriggerResultsBasedEventSelector.h]: @Dr15Jones No, it should not be needed. However, the optimal fix to get rid of it invlves other changes, and will be done in a separate PR, to avoid having to rebuild > 800 packages when testing.

**Comment by wmtan** [comment]
Don't do anything with this PR yet. There is another update coming.

**Comment by slava77** [comment]
@wmtan is that update still coming soon?

**Comment by wmtan** [comment]
@slava77 Coincidentally, the update came two minutes after your query.

**Comment by wmtan** [comment]
@Dr15Jones This extensive update addresses all of your comments, except for the one about eliminating the old NamedEventSelector ctor, which we agreed I would do in a separate PR. The update also fixes a pre-existing bug in SubProcess.cc (missing setComplete() calls) that was discovered while testing this update.

**Comment by wmtan** [comment]
@Dr15Jones @slava77 
Hold it again. I just noticed that two new files that I added in my latest commit do not appear in this PR, even though they show as committed in my working node.
So, there is some git wierdness going on that I do not fully understand yet.
I will have to push again after this is fixed.

**Comment by wmtan** [comment]
@Dr15Jones @slava77 Should be OK now. I had forgotten the "git add" for two new files.

**Comment by wmtan** [comment]
@cmsbuild please test

**Comment by wmtan** [comment]
@Dr15Jones I extensively rewrote the RootOutputFile::fillBranches() function, not only to use the new base class for the "ForOutput" class, but to make the function more concise and understandable. I think it is a great improvement.

**Comment by wmtan** [comment]
@cmsbuild @Dr15Jones The test failures appear to have nothing to do with this PR. The runtestTqafTopEventSelection test fails in the base IB. The TestRunnerPhysicsToolsCondLiteIO test runs successfully in my work area with this PR, although I am running on a slightly older IB.
Also, the error in TestRunnerPhysicsToolsCondLiteIO appears to have no connection to anything I changed. I will investigate further, however, just to be safe.

**Comment by wmtan** [comment]
@cmsbuild @Dr15Jones OK, I ran the failing TestRunnerPhysicsToolsCondLiteIO with this PR against the latest IB, CMSSW_8_1_X_2016-03-09-1100, and the test passed. I cannot reproduce this test failure. So neither unit test failure is related to this PR.

File: IOPool/Output/src/RootOutputFile.cc [hunk 5]
```diff
@@ -682,70 +684,53 @@ namespace edm {
...
-    ProductProvenanceRetriever const* provRetriever = nullptr;
-    {
-      EventPrincipal const* eventPrincipal = dynamic_cast<EventPrincipal const*>(&principal);
-      if (eventPrincipal) {
-        provRetriever =eventPrincipal->productProvenanceRetrieverPtr();
-        assert(provRetriever);
-      }
-    }
-    
+    bool const fastCloning = (branchType == InEvent) && (whyNotFastClonable_ == FileBlock::CanFastClone);
```
Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputFile.cc]: We can only fast clone events?

Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputFile.cc]: Ah, I see that was in the original logic.

File: IOPool/Output/src/RootOutputFile.cc [hunk 6]
```diff
@@ -682,70 +684,53 @@ namespace edm {
...
       branchesWithStoredHistory_.insert(id);
 
       bool produced = item.branchDescription_->produced();
-      bool keepProvenance = productProvenanceVecPtr != nullptr &&
-                            (om_->dropMetaData() == PoolOutputModule::DropNone ||
-                             om_->dropMetaData() == PoolOutputModule::DropDroppedPrior ||
-                            (om_->dropMetaData() == PoolOutputModule::DropPrior && produced));
-      bool getProd = (produced || !fastCloning ||
-         treePointers_[branchType]->uncloned(item.branchDescription_->branchName()));
+      bool getProd = (produced || !fastCloning || treePointers_[InEvent]->uncloned(item.branchDescription_->branchName()));
```
Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputFile.cc]: There is no guarantee that the item you are looking at is InEvent.

File: IOPool/Output/src/RootOutputFile.cc [hunk 7]
```diff
@@ -682,70 +684,53 @@ namespace edm {
...
       if(getProd) {
-        OutputHandle const oh = principal.getForOutput(id, mcc);
-        product = oh.wrapper();
-        productProvenance = oh.productProvenance();
-      } else {
-        if(keepProvenance) {
-          productProvenance = provRetriever->branchIDToProvenance(id);
+        bool found = occurrence.getByToken(item.token_, item.branchDescription_->unwrappedTypeID(), result);
+        product = result.wrapper();
+        if(found && keepProvenance) {
```
Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputFile.cc]: Couldn't the product not be found but the provenance still be there? This would be the case where we did not write out the product but kept the provenance in a previous step.

Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputFile.cc]: Ah, I see later that there is a separate check which handles that case.

**Comment by Dr15Jones** [comment]
Done with next review

File: IOPool/Output/src/RootOutputFile.cc (hunk 6 above)
Comment (wmtan) [inline IOPool/Output/src/RootOutputFile.cc]: @Dr15Jones Actually, there currently is such a guarantee, because if it is not InEvent, fastCloning will always be false and we never get there. However, the guarantee is based on an assumption that may not always be true in the future. So, the code would be more maintainable and more future proof if InEvent were replaced by branchType here, so I will do so,

File: IOPool/Output/src/RootOutputFile.cc (hunk 7 above)
Comment (wmtan) [inline IOPool/Output/src/RootOutputFile.cc]: @Dr15Jones Yes, it's the same separate check as before, but it got moved down a few lines, so you hadn't gotten there yet.

File: IOPool/Output/src/RootOutputFile.cc (hunk 5 above)
Comment (wmtan) [inline IOPool/Output/src/RootOutputFile.cc]: @Dr15Jones A decision was made a long time ago to never fast copy anything in the run or lumi trees. It makes life a lot simpler.

**Comment by wmtan** [comment]
Incorporated latest comments from Chris Jones. Should be all done now.

**Comment by wmtan** [comment]
@cmsbuild please test
